### PR TITLE
tests: rewrite gpg-agent shutdown code

### PR DESCRIPTION
**Pull request by petermax2**
to use fork () and execv () instead of system ().
See #2341 for full discussion.

## Basics

These points need to be fulfilled for every PR:

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy.
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

**Comment by petermax2** [comment]
GPG also causes memory leaks on Debian unstable during decryption. Never seen this before:

    ==19935== Command: /usr/bin/gpg2 --batch --trust-model always -d

**Comment by petermax2** [comment]
There seems to be a problem with the Ruby bindings on Cirrus:

	FAILED: src/plugins/ruby/CMakeFiles/elektra-ruby-objects.dir/ruby.cpp.o 
	/usr/local/opt/llvm/bin/clang++  -DELEKTRA_MODULE_NAME=ruby -DELEKTRA_PLUGIN_NAME=\"ruby\" -DELEKTRA_PLUGIN_NAME_C=ruby -DELEKTRA_STATIC -DHAVE_KDBCONFIG_H -DPLUGIN_SHORT_NAME=ruby -DREADME=readme_ruby.c -DSWIG_RUNTIME=\"runtime.h\" -DSWIG_TYPE_TABLE=kdb -Isrc/include -I../src/include -Isrc/bindings/cpp/include -I../src/bindings/cpp/include -Isrc/plugins/ruby -std=c++11  -Wno-deprecated-declarations  -Wold-style-cast -Wno-missing-field-initializers -Woverloaded-virtual  -Wsign-promo -Werror -Wno-long-long -Wpedantic -Wno-variadic-macros -Wall -Wextra -Wno-overlength-strings -Wsign-compare -Wfloat-equal -Wformat -Wformat-security -Wshadow -Wcomments -Wtrigraphs -Wundef -Wuninitialized -Winit-self -O2 -g -DNDEBUG -isysroot /Applications/Xcode-11.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk   -fPIC -isystem  "/usr/local/Cellar/ruby/2.7.1_2/include/ruby-2.7.0" -isystem  "/usr/local/Cellar/ruby/2.7.1_2/include/ruby-2.7.0/x86_64-darwin19" -Wno-shadow -Wno-old-style-cast -Wno-unused-variable -Wno-unused-parameter -Wno-deprecated-register -Wno-macro-redefined -Wno-reserved-user-defined-literal -Wno-gnu-zero-variadic-macro-arguments -MD -MT src/plugins/ruby/CMakeFiles/elektra-ruby-objects.dir/ruby.cpp.o -MF src/plugins/ruby/CMakeFiles/elektra-ruby-objects.dir/ruby.cpp.o.d -o src/plugins/ruby/CMakeFiles/elektra-ruby-objects.dir/ruby.cpp.o -c ../src/plugins/ruby/ruby.cpp
	In file included from ../src/plugins/ruby/ruby.cpp:22:
	src/plugins/ruby/runtime.h:1108:3: error: call to 'rb_define_virtual_variable' is ambiguous
	  rb_define_virtual_variable("SWIG_TRACKINGS_COUNT", swig_ruby_trackings_count, NULL);
	  ^~~~~~~~~~~~~~~~~~~~~~~~~~
	/usr/local/Cellar/ruby/2.7.1_2/include/ruby-2.7.0/ruby/backward/cxxanyargs.hpp:59:1: note: candidate function
	rb_define_virtual_variable(const char *q, type *w, void_type *e)
	^
	/usr/local/Cellar/ruby/2.7.1_2/include/ruby-2.7.0/ruby/backward/cxxanyargs.hpp:90:1: note: candidate function
	rb_define_virtual_variable(const char *q, type *w, rb_gvar_setter_t *e)
	^
	1 error generated.

**Comment by mpranj** [comment]
> There seems to be a problem with the Ruby bindings on Cirrus:

Thank you for reporting. This is due to swig being incompatible with ruby 2.7. We already encountered it, but it's unclear why this appeared on Cirrus just now.

**Comment by petermax2** [comment]
Should I open a new issue for it?

**Comment by mpranj** [comment]
It's already reported (#3378). We're basically waiting on swig to make a release. In the mean time I will downgrade the ruby version on cirrus macOS build jobs.

**Comment by petermax2** [comment]
grafik
I hate to tell you.... but it works inside my container :cry: :sob: I used the same image as Jenkins and copied the build instructions from Jenkin's logs.

**Comment by mpranj** [comment]
@petermax2 did you download the exact image from our hub or re-build it yourself?  Package versions are not pinned so maybe you have a newer version of some package if you build the image yourself (i.e. the bug was patched). 

Does this error require the "high load" scenario or is that irrelevant here?

Btw since when does the memory check scramble the output. This is terrible to read. :fearful:

**Comment by petermax2** [comment]
> @petermax2 did you download the exact image from our hub or re-build it yourself? Package versions are not pinned so maybe you have a newer version of some package if you build the image yourself (i.e. the bug was patched).

I pulled the image 

    podman pull hub-public.libelektra.org/build-elektra-debian-sid:202004-72b60448c4719fea1836b0e7441abfb5016694fb04025a8c0b1f7fe12971193c

> 
> Does this error require the "high load" scenario or is that irrelevant here?
> 

Well our old bug (failed to close gpg-agent) required some sort of "high load" scenario. But here we have repeatedly failing test cases. The test cases only fail on the build server and not locally.

> Btw since when does the memory check scramble the output. This is terrible to read. fearful

This is really strange and I don't know what's happening on the build server.

**Comment by mpranj** [comment]
A similar (the same) memory leak now [appeared in my PR](https://build.libelektra.org/job/libelektra/job/PR-3411/6/artifact/artifacts/debian-unstable-full-clang/DynamicAnalysis.xml) as well, with no changes to the crypto plugin code.

**Comment by petermax2** [comment]
Maybe the build server needs a restart?

**Comment by mpranj** [comment]
Notably it also appeared on [Fedora 31](https://build.libelektra.org/job/libelektra/job/PR-3411/7/artifact/artifacts/fedora-31-full/DynamicAnalysis.xml), not on the same buildserver.

> Maybe the build server needs a restart?

Not sure why this would help. It occured on separate build servers with different hardware. All of them are regularly patched and rebooted.

**Comment by petermax2** [comment]
I came up with another hypothesis: I think we should set the environment variable `GNUPGHOME` to an empty temporary directory before each test run on the buildserver. Maybe if a build job is cancelled or if it fails, there is no proper clean-up in the `GNUPGHOME` dir.

Where would you put the commands to create a temporary directory and use it as `GNUPGHOME` in the `Jenkinsfile`? Is the `ctest` method a good place for this?

File: scripts/jenkins/Jenkinsfile [hunk 1]
```diff
@@ -1305,9 +1305,12 @@ def ctest(target = "Test") {
      , since they are known to fail in scenarios with high load.
```
Comment (markus2330) [inline scripts/jenkins/Jenkinsfile:1305]: The text above does not seem to match what is done below.

**Comment by markus2330** [comment]
We already do similar things in tests/cframework/tests.c. Maybe you can solve a more general problem: that all things from our test runs get cleaned up at the end. Otherwise, temporary test failures should not be a problem anymore as we re run failed tests anyway (at least that happens in scripts/dev/run_all.in, I wonder why Jenkins has its own logic to start the tests?). In any case: think big how to make the test framework generally more stable :+1:

**Comment by mpranj** [comment]
> there is no proper clean-up in the GNUPGHOME dir.

Is this relevant as we spawn a new container for each run?

**Comment by mpranj** [comment]
In #3411 I disabled crypto on Fedora 31 and Debian unstable and made other fixes to make master builds green. I suggest we merge that and then enable crypto in this PR again to debug the problems.

**Comment by petermax2** [comment]
> In #3411 I disabled crypto on Fedora 31 and Debian unstable and made other fixes to make master builds green. I suggest we merge that and then enable crypto in this PR again to debug the problems.

I agree!

**Comment by markus2330** [comment]
I also agree with @mpranj approach to first make green, and debug later (probably with a cleaner approach how we do cleanup in general).

> Is this relevant as we spawn a new container for each run?

In the case of rerunning failed test cases it might be relevant.

**Comment by mpranj** [comment]
> In the case of rerunning failed test cases it might be relevant.

I did not think of that, but afaik our jenkins does not re-run tests as it does not use the `run_all` script.

**Comment by markus2330** [comment]
> our jenkins does not re-run tests as it does not use the run_all script.

I created #3414 for further discussions on this topic.

**Comment by mpranj** [comment]
I just noticed another (sporadic?) error during an [unrelated PR](https://build.libelektra.org/blue/organizations/jenkins/libelektra/detail/PR-3413/2/pipeline). @petermax2 as far as I see this is the exact code you're trying to improve within this PR #3410?

```
128/137 MemCheck  #60: testmod_crypto ...................***Failed   32.49 sec

OK closing connection

CRYPTO       TESTS

==================



/home/jenkins/workspace/libelektra_PR-3413/src/plugins/crypto/gpgagent_teardown.h:16: warn in test_teardown: failed to stop the gpg-agent, trying to kill it

crypto Results: 67 Tests done — 0 errors.
```

**Comment by petermax2** [comment]
> @petermax2 as far as I see this is the exact code you're trying to improve within this PR #3410?

Yes, exactly!

**Comment by petermax2** [comment]
Thanks to @mpranj we have a green build on Jenkins here. I will re-run the test a view times to check if the memory leak occurs again during gpg-agent shutdown.

**Comment by mpranj** [comment]
@petermax2 I did not fix anything. I just disabled the problematic tests so they do not bother others.

**Comment by petermax2** [comment]
I'll just rebase and re-enable them.

**Comment by petermax2** [comment]
With the docker image I am able to reproduce the memory leaks (now I see more leaks locally than on the server):

    podman pull hub-public.libelektra.org/build-elektra-debian-sid:202004-72b60448c4719fea1836b0e7441abfb5016694fb04025a8c0b1f7fe12971193c
    podman run -t -i 556e8ee5ff6c /bin/bash

Following the instructions of the build server (see [log](https://build.libelektra.org/blue/rest/organizations/jenkins/pipelines/libelektra/branches/PR-3410/runs/9/nodes/654/log/?start=0&download=true)).

:tada: 

The memory leaks all occur because of `gpg`. Even for `fcrypt` leaks are detected. Seems like there are memory leaks in `gpg` on Debian Sid. I also got the "chopped" output of the `gpgme` unit test:

	Command: "/usr/bin/valgrind" "--log-file=/home/jenkins/workspace/libelektra_PR-3410/build/Testing/Temporary/MemoryChecker.77.log" "--trace-children=yes" "--leak-check=full" "--gen-suppressions=all" "--error-exitcode=1" "--suppressions=/home/jenkins/workspace/libelektra_PR-3410/tests/valgrind.suppression" "/home/jenkins/workspace/libelektra_PR-3410/build/bin/testmod_gpgme" "/home/jenkins/workspace/libelektra_PR-3410/src/plugins/gpgme"
	Directory: /home/jenkins/workspace/libelektra_PR-3410/src/plugins/gpgme
	"testmod_gpgme" start time: Apr 29 16:15 UTC
	Output:
	----------------------------------------------------------
	GPGME        TESTS
	==================

	GPGME        TESTS
	==================

	GPGME        TESTS
	==================

	GPGME        TESTS
	==================

	GPGME        TESTS
	==================

	GPGME        TESTS
	==================

	GPGME        TESTS
	==================

	GPGME        TESTS
	==================

	GPGME        TESTS
	==================

	GPGME        TESTS
	==================

	GPGME        TESTS
	==================

	GPGME        TESTS
	==================

	GPGME        TESTS
	==================


	gpgme Results: 45 Tests done — 0 errors.
	<end of output>
	Test time =  12.16 sec
	----------------------------------------------------------
	Test Passed.
	"testmod_gpgme" end time: Apr 29 16:15 UTC
	"testmod_gpgme" time elapsed: 00:00:12
	----------------------------------------------------------

A good way of triggering the error is to run all crypto/gpg-ish tests in parallel like:

    ctest -j 3 --force-new-ctest-process --output-on-failure --no-compress-output -T MemCheck -LE memleak -R "(crypto|fcrypt|gpgme)"

**Comment by petermax2** [comment]
This PR is hopefully an improvement to the situation but maybe not the solution. We still have to keep an eye on the unit tests.

`gpg` definitely leaks during decryption (even when using `libgpgme`) on Debian Stable and Fedora 31. If we invoke the memchecker with `--trace-children=yes` we will always detect the memory leak of `gpg` right after the gpg module forks.

@mpranj @markus2330 please review my changes. What's your opinion about the memory leak?

File: scripts/jenkins/Jenkinsfile [hunk 2]
```diff
@@ -587,7 +587,7 @@ def generateFullBuildStages() {
     CMAKE_FLAGS_BUILD_ALL +
       CMAKE_FLAGS_DEBUG +
       CMAKE_FLAGS_CLANG + [
-      'PLUGINS': 'ALL;-crypto;-fcrypt',
+      'PLUGINS': 'ALL;-crypto;-fcrypt;-gpgme',
```
Comment (mpranj) [inline scripts/jenkins/Jenkinsfile:590]: I see a different diff for `Jenkinsfile` locally? :confused: Do you need to rebase or is this a github problem?

Comment (mpranj) [reply scripts/jenkins/Jenkinsfile:590]: Actually, the question is: can these tests be enabled again for `debian-unstable-full-clang` or do they still cause problems?

Comment (petermax2) [reply scripts/jenkins/Jenkinsfile:590]: Ah you re-enabled the unit tests :-) Should we leave them enabled or not?

Comment (petermax2) [reply scripts/jenkins/Jenkinsfile:590]: Let's try them, but I think they still caused problems the last time.

Comment (mpranj) [reply scripts/jenkins/Jenkinsfile]: Thank you for reacting so quickly. You can also leave some tests disabled, but please open another PR where they are enabled (or an issue?), so we do not forget that it needs to be fixed.

**Comment by petermax2** [comment]
> I'm just baffled that replacing the `system()` call with a basically equivalent `execv()` call improves things. On linux the implementation of `system()` even calls `fork()` and `execl()`. confused

To be honest, I have no clue why this would improve anything. My first idea was that maybe fork fails during high server load. Also it took me a long time to even reproduce the leak locally within a container.

**Comment by petermax2** [comment]
The second build run was also green. I'm merging now.

**Pull request merged**